devth/yetibot
# Google custom search (images and text) #524

## Goal

To create  `google` commands that is able to search both text and images. 
Users should also be able to customize the search engine via the config file.
User should  be able to put in cli type options while using the `search` and `image` (image search command)
User should be able to set options for search via the appropriate commands 

## Accompilshed
- `google search <query>` returns at-most 10 results (because of custom engine API limitation) with a `title`,`link` and `snippet` for each result.
-   `google image <query>` now searches for images based on what the query is. The image results are displayed in a different order than the vanilla search one due to rendering weirdness on slack
- config is checked for the existence of two keys in existence in addition  to the config-validation function  that exists in `yeti.core` 
- messages are displayed to the user if no results are found
- warning messages are displayed on the server (where the yetibot is running) console in case google returns messages with http failure status codes.
- cli type options can now be put in `search` and `image` commands 
- options can be set via the config file or (inclusive) via the commands on the chat frontend.
- options can be completely rest via the commandline or (inclusive) also be restored to the existing values set in the config file, again via the commands on the frontend.
- options that have been set can be viewed on the chat frontend

## how to test
1. run yetibot locally and have it access slack correctly
2. remove google search engine credentials, restart yetibot and check to see if it logs out that `Google is not configured`
3. Add-in one of the two keys-value pairs ,restart yetibot and verify you see that the message `Google is not configured` is outputted to the console
4. Put in incorrect credentials (for api-key and custom-search-engine-id) and run yetibot and try to enter something like `!google search digimon` (on slack) and you should given the `API credentials are invalid || google died` message. Also check your console for info messages.
5. Put in the correct credentials but keep the `image search` setting disabled and restart yetibot, this time `!google search digimon` should give you results but `!google image digimon` should give you the message `Google image search returned no results!! ... check your search engine settings` 
6. Once you enable the image search part, a query like `!google image ehhdfskjsfdfsdjfhkjh87dhfjksdhjkds` should return the same message as mentioned above ("Google image returned ..... please...)
7. `!google search fshsfdhfshsfdjh8978978sfdjksfdhkh` should return the message `"Google returned no results!!"`
8. Put in in-valid option values in the config file and restart yetibot
9. Validation errors should appear on the terminal output while the yetibot starts up
10. putting in `google show set options` should show `No options have been set`
11. for a list of options that can be set via the chat frontend enter in `google show options`
12. Try getting info on the commands that you can see as a result of entering the above command
13. Try setting options via the `!google set <option> <value>`
14. Try to make sure sure your search results reflect the options you have set
15. Reset these options by entering the command `!google reset options`
16. Now try entering options via option arguments to the search functions and verify the correctness of the search results. 

## for credentials
1. Have a google/gmail account
2. go here https://cse.google.com/cse/all
